### PR TITLE
[Merged by Bors] - Document service discovery

### PR DESCRIPTION
**Pull request by maltesander**
## Description

closes https://github.com/stackabletech/hive-operator/issues/132

## Review Checklist
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added (or not applicable)
- [ ] Documentation added (or not applicable)
- [ ] Changelog updated (or not applicable)
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
- [ ] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: docs/modules/ROOT/pages/discovery.adoc [hunk 1]
```diff
@@ -0,0 +1,50 @@
+:clusterName: simple-hive-derby
+:namespace: stackable
+:metastorePort: 9083
```
Comment (teozkr) [inline docs/modules/ROOT/pages/discovery.adoc:3]: Can we use some more obvious placeholder names?

Comment (maltesander) [reply docs/modules/ROOT/pages/discovery.adoc:3]: like prefixing with `hive` or what do you suggest?

Comment (teozkr) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Using something like `$CLUSTERNAME` as the cluster name.

Comment (teozkr) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Looks like it works without generating extra warnings if I declare `:clusterName: \{clusterName\}` (and so on).

Comment (soenkeliebau) [reply docs/modules/ROOT/pages/discovery.adoc:3]: I'm not sure if I'm following the discussion correctly here, is it about if the exapmle should be this:
```
apiVersion: hive.stackable.tech/v1alpha1
kind: HiveCluster
metadata:
  name: {clusterName} # <1>
  namespace: {namespace} # <2>
spec:
...
```

vs this:

```
apiVersion: hive.stackable.tech/v1alpha1
kind: HiveCluster
metadata:
  name: simple-hive # <1>
  namespace: default # <2>
spec:
  ...
```

Comment (teozkr) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Yeah, but mostly in the context of the discovery CMs, not the cluster definition itself.

Comment (soenkeliebau) [reply docs/modules/ROOT/pages/discovery.adoc:3]: So [here](https://github.com/stackabletech/hive-operator/pull/145/files#diff-b83490b212c5dfb5449c7b6df704ebe41c33ed4fe36223a1df2ad7399b968bb1R37) ?

Comment (teozkr) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Yep, and the URLs listed below that.

Comment (maltesander) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Yeah if we use the abstract names we should adapt the examples of course as well. I like it explicit (and thats how it is in opa and kafka as well).

Comment (teozkr) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Err, the abstract version is explicit about where the values are coming from. The _concrete_ version only gives a weak implication about the relationship.

Comment (soenkeliebau) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Can we do something like this as a compromise: 

> The resulting discovery ConfigMap follows the naming convention  `{namespace}/{clusterName}`, which for the example shown above would resolve to 'stackable/simple-hive-derby' ...

Comment (maltesander) [reply docs/modules/ROOT/pages/discovery.adoc:3]: I think Teo refers to the explicit names in the connection strings:
```
`HIVE`::
====
Contains the thrift protocol connection string for the Hive metastore:
[subs="attributes"]
  thrift://{clusterName}-metastore-{roleGroup}-0.{clusterName}-metastore-{roleGroup}.{namespace}.svc.cluster.local:{metastorePort}
```
should stay as is (without the replacements)

Comment (soenkeliebau) [reply docs/modules/ROOT/pages/discovery.adoc:3]: Ah! Sorry for being daft today (and others days as well I guess).

I think it is better with the abstract placeholders in there, yes. Same idea if we wanted to make it explicit: generate a concrete example on top of the abstract version, but if we have to pick one of the two I think it should be the abstract one.

File: docs/modules/ROOT/pages/discovery.adoc [hunk 2]
```diff
@@ -0,0 +1,53 @@
+:clusterName: \{clusterName\}
+:namespace: \{namespace\}
+:metastorePort: 9083
+:roleGroup: \{roleGroup\}
+
+= Discovery
+
+The Stackable Operator for Apache Hive publishes a discovery https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#configmap-v1-core[`ConfigMap`], which exposes a client configuration bundle that allows access to the Apache Hive cluster.
+
+The bundle includes a thrift connection string to access the Hive metastore service. This string may be used by other operators or tools to configure their products with access to Hive. This is limited to internal cluster access.
```
Comment (teozkr) [inline docs/modules/ROOT/pages/discovery.adoc:10]:
```suggestion
The bundle includes an Apache Thrift connection string to access the Hive Metastore service. This string may be used by other operators or tools to configure their products with access to Hive. Access is limited to services within the same Kubernetes cluster.
```

File: docs/modules/ROOT/pages/discovery.adoc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+Contains the thrift protocol connection string for the Hive metastore:
+
+[subs="attributes"]
+  thrift://{clusterName}-metastore-{roleGroup}-0.{clusterName}-metastore-{roleGroup}.{namespace}.svc.cluster.local:{metastorePort}
+
+or
+
+[subs="attributes"]
+  thrift://{clusterName}-metastore-{roleGroup}-0.{clusterName}-metastore-{roleGroup}.{namespace}.svc.cluster.local:{metastorePort}
+  thrift://{clusterName}-metastore-{roleGroup}-1.{clusterName}-metastore-{roleGroup}.{namespace}.svc.cluster.local:{metastorePort}
```
Comment (teozkr) [inline docs/modules/ROOT/pages/discovery.adoc:50]: Raising this here anyway, but this is more of an implementation question that we might want to raise as a separate issue...

Is this format for failover actually supported by Hive/Thrift clients? What are the consequences here if running in HA mode without upstream support? (Stale caches, data corruption, ???) Depending on the answer here, _do we want to support this mode of operation_?

**Review by teozkr — APPROVED** [review]

File: docs/modules/ROOT/pages/discovery.adoc (hunk 3 above)
Comment (maltesander) [reply docs/modules/ROOT/pages/discovery.adoc:50]: We do not have HA. These would be completely separated metastore instances afaik (there fore no caches or data corruption). I thought about limiting the replicas to 1 already but never introduced that. 

So yeah i would not support that for now.

**Comment by maltesander** [comment]
bors merge

**Pull request closed without merging**